src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/processor/EnvelopeProcessor.java

```diff
@@ -53,28 +53,37 @@ public Envelope processEnvelope(byte[] metadataStream) throws IOException {
     }
 
     public void markAsUploaded(Envelope envelope, String containerName, String zipFileName) {
-        EnvelopeState state = new EnvelopeState(containerName, zipFileName, DOC_UPLOADED);
+        ProcessEvent event = new ProcessEvent(containerName, zipFileName, DOC_UPLOADED);
 
-        state.setEnvelope(envelope);
+        event.setEnvelope(envelope);
 
-        envelopeStateRepository.save(state);
+        processEventRepository.save(event);
+
+        envelope.setLastEvent(DOC_UPLOADED);
+        envelopeRepository.save(envelope);
     }
 
     public void markAsUploadFailed(String reason, Envelope envelope, String container, String zipFileName) {
-        EnvelopeState state = new EnvelopeState(container, zipFileName, DOC_UPLOAD_FAILURE);
+        ProcessEvent event = new ProcessEvent(container, zipFileName, DOC_UPLOAD_FAILURE);
+
+        event.setEnvelope(envelope);
+        event.setReason(reason);
 
-        state.setEnvelope(envelope);
-        state.setReason(reason);
+        processEventRepository.save(event);
 
-        envelopeStateRepository.save(state);
+        envelope.setLastEvent(DOC_UPLOAD_FAILURE);
```

Comment:
If last state in envelope table for a envelope is `DOC_UPLOAD_FAILURE` then how would the Job scheduler task know how many times it needs to retry as envelope table doesn't have a date from which task can derive when it was first created.

Reply:
As you mentioned in above comment there would be createdAt column in envelope so it should be derive the logic.